common/src/main/java/io/druid/math/expr/Function.java

```diff
@@ -29,6 +31,10 @@
 
   Number apply(List<Expr> args, Expr.ObjectBinding bindings);
 
+  // optional interface to be used when function should be created per reference in expression
+  interface FunctionFactory extends Supplier<Function>, Function {
```

Comment:
Did this not to make two classes (factory and function) for each stateful functions. if it's look that bad, I can separate these two.